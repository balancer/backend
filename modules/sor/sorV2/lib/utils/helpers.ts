import {
    BatchSwapStep,
    BigintIsh,
    DEFAULT_USERDATA,
    PriceImpactAmount,
    SingleSwap,
    SwapKind,
    Token,
    TokenAmount,
} from '@balancer/sdk';
import { PathWithAmount } from '../path';
import { MathSol, abs } from './math';

export function checkInputs(
    tokenIn: Token,
    tokenOut: Token,
    swapKind: SwapKind,
    swapAmount: BigintIsh | TokenAmount,
): TokenAmount {
    let amount: TokenAmount;

    if (swapAmount instanceof TokenAmount) {
        amount = swapAmount;
    } else {
        amount = TokenAmount.fromRawAmount(swapKind === SwapKind.GivenIn ? tokenIn : tokenOut, swapAmount);
    }

    if (
        (swapKind === SwapKind.GivenIn && !tokenIn.isEqual(amount.token)) ||
        (swapKind === SwapKind.GivenOut && !tokenOut.isEqual(amount.token))
    ) {
        throw new Error('Swap amount token does not match input token');
    }

    return amount;
}

export function calculatePriceImpact(paths: PathWithAmount[], swapKind: SwapKind): PriceImpactAmount {
    const pathsReverse = paths.map(
        (path) =>
            new PathWithAmount(
                [...path.tokens].reverse(),
                [...path.pools].reverse(),
                [...path.operations].reverse(),
                swapKind === SwapKind.GivenIn ? path.outputAmount : path.inputAmount,
            ),
    );

    const amountInitial =
        swapKind === SwapKind.GivenIn ? getInputAmount(paths).amount : getInputAmount(pathsReverse).amount;

    const amountFinal =
        swapKind === SwapKind.GivenIn ? getOutputAmount(pathsReverse).amount : getOutputAmount(paths).amount;

    const priceImpact = MathSol.divDownFixed(amountInitial - amountFinal, amountInitial * 2n);
    return PriceImpactAmount.fromRawAmount(priceImpact);
}

export function getInputAmount(paths: PathWithAmount[]): TokenAmount {
    if (!paths.every((p) => p.inputAmount.token.isEqual(paths[0].inputAmount.token))) {
        throw new Error('Input amount can only be calculated if all paths have the same input token');
    }
    const amounts = paths.map((path) => path.inputAmount);
    return amounts.reduce((a, b) => a.add(b));
}

export function getOutputAmount(paths: PathWithAmount[]): TokenAmount {
    if (!paths.every((p) => p.outputAmount.token.isEqual(paths[0].outputAmount.token))) {
        throw new Error('Output amount can only be calculated if all paths have the same output token');
    }
    const amounts = paths.map((path) => path.outputAmount);
    return amounts.reduce((a, b) => a.add(b));
}
