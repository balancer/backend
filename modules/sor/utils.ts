import { tokenService } from '../token/token.service';
import { Chain } from '@prisma/client';
import { AllNetworkConfigsKeyedOnChain, chainToIdMap } from '../network/network-config';
import { GqlSorGetSwapPaths, GqlSorGetSwapsResponse, GqlSorSwapType } from '../../schema';
import { replaceZeroAddressWithEth } from '../web3/addresses';
import { Address } from 'viem';
import { Token, TokenAmount } from '@balancer/sdk';

export async function getTokenAmountHuman(tokenAddr: string, humanAmount: string, chain: Chain): Promise<TokenAmount> {
    const token = await getToken(tokenAddr, chain);
    return TokenAmount.fromHumanAmount(token, humanAmount as `${number}`);
}

export async function getTokenAmountRaw(tokenAddr: string, rawAmount: string, chain: Chain): Promise<TokenAmount> {
    const token = await getToken(tokenAddr, chain);
    return TokenAmount.fromRawAmount(token, rawAmount);
}

/**
 * Gets a b-sdk Token based off tokenAddr.
 * @param address
 * @param chain
 * @returns
 */
export const getToken = async (tokenAddr: string, chain: Chain): Promise<Token> => {
    if (tokenAddr === AllNetworkConfigsKeyedOnChain[chain].data.eth.address) {
        return new Token(
            parseFloat(chainToIdMap[chain]),
            AllNetworkConfigsKeyedOnChain[chain].data.weth.address as Address,
            18,
        );
    } else {
        const prismaToken = await tokenService.getToken(tokenAddr, chain);
        if (!prismaToken) throw Error(`Missing token from tokenService ${tokenAddr}`);
        return new Token(parseFloat(chainToIdMap[chain]), prismaToken.address as Address, prismaToken.decimals);
    }
};

export const swapPathsZeroResponse = (tokenIn: string, tokenOut: string): GqlSorGetSwapPaths => {
    return {
        swaps: [],
        paths: [],
        tokenAddresses: [],
        swapType: 'EXACT_IN',
        vaultVersion: 2,
        tokenIn: replaceZeroAddressWithEth(tokenIn),
        tokenOut: replaceZeroAddressWithEth(tokenOut),
        tokenInAmount: '0',
        tokenOutAmount: '0',
        swapAmount: '0',
        swapAmountScaled: '0',
        returnAmount: '0',
        returnAmountScaled: '0',
        effectivePrice: '0',
        effectivePriceReversed: '0',
        routes: [],
        priceImpact: '0',
    };
};

export const zeroResponse = (
    swapType: GqlSorSwapType,
    tokenIn: string,
    tokenOut: string,
    swapAmount: string,
): GqlSorGetSwapsResponse => {
    return {
        marketSp: '0',
        tokenAddresses: [],
        swaps: [],
        tokenIn: replaceZeroAddressWithEth(tokenIn),
        tokenOut: replaceZeroAddressWithEth(tokenOut),
        swapType,
        tokenInAmount: swapType === 'EXACT_IN' ? swapAmount : '0',
        tokenOutAmount: swapType === 'EXACT_IN' ? '0' : swapAmount,
        swapAmount: swapType === 'EXACT_IN' ? '0' : swapAmount,
        swapAmountScaled: '0',
        swapAmountForSwaps: '0',
        returnAmount: '0',
        returnAmountScaled: '0',
        returnAmountConsideringFees: '0',
        returnAmountFromSwaps: '0',
        routes: [],
        effectivePrice: '0',
        effectivePriceReversed: '0',
        priceImpact: '0',
    };
};
