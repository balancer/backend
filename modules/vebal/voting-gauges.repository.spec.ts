import { setMainnetRpcProviderForTesting } from '../../test/utils';
import { defaultStakingGaugeId, prismaMock } from './prismaPoolStakingGauge.mock';
import { VotingGauge, VotingGaugesRepository } from './voting-gauges.repository';
import { Chain } from '@prisma/client';

const httpRpc = 'http://127.0.0.1:8555';
setMainnetRpcProviderForTesting(httpRpc);

it('maps onchain network format into prisma chain format', async () => {
    const repository = new VotingGaugesRepository();
    expect(repository.toPrismaNetwork('Mainnet')).toBe(Chain.MAINNET);
    expect(repository.toPrismaNetwork('Optimism')).toBe(Chain.OPTIMISM);
    expect(repository.toPrismaNetwork('veBAL')).toBe(Chain.MAINNET);
    expect(repository.toPrismaNetwork('POLYGONZKEVM')).toBe(Chain.ZKEVM);
    expect(() => repository.toPrismaNetwork('Unknown')).toThrowError('Network UNKNOWN is not supported');
});

it('fetches list of voting gauge addresses', async () => {
    const repository = new VotingGaugesRepository();
    const addresses = await repository.getVotingGaugeAddresses();
    expect(addresses.length).toBe(373);
}, 10_000);

it('generates voting gauge rows given a list of gauge addresses', async () => {
    const repository = new VotingGaugesRepository();

    const votingGaugeAddresses = [
        '0x79eF6103A513951a3b25743DB509E267685726B7',
        '0xfb0265841C49A6b19D70055E596b212B0dA3f606',
        '0x8F7a0F9cf545DB78BF5120D3DBea7DE9c6220c10',
    ];
    // Uncomment to test with all the root gauges
    // const votingGaugeAddresses = await service.getVotingGaugeAddresses();

    const rows = await repository.fetchOnchainVotingGauges(votingGaugeAddresses);

    expect(rows).toMatchInlineSnapshot(`
      [
        {
          "gaugeAddress": "0x79ef6103a513951a3b25743db509e267685726b7",
          "isInSubgraph": false,
          "isKilled": false,
          "network": "MAINNET",
          "relativeWeight": 0.07308250729037725,
          "relativeWeightCap": undefined,
        },
        {
          "gaugeAddress": "0xfb0265841c49a6b19d70055e596b212b0da3f606",
          "isInSubgraph": false,
          "isKilled": true,
          "network": "OPTIMISM",
          "relativeWeight": 0,
          "relativeWeightCap": undefined,
        },
        {
          "gaugeAddress": "0x8f7a0f9cf545db78bf5120d3dbea7de9c6220c10",
          "isInSubgraph": false,
          "isKilled": false,
          "network": "ARBITRUM",
          "relativeWeight": 0,
          "relativeWeightCap": "0.02",
        },
      ]
    `);
}, 10_000);

it('Excludes Liquidity Mining Committee gauge', async () => {
    const liquidityMiningAddress = '0x7AA5475b2eA29a9F4a1B9Cf1cB72512D1B4Ab75e';
    const repository = new VotingGaugesRepository();
    const rows = await repository.fetchOnchainVotingGauges([liquidityMiningAddress]);
    expect(rows).toEqual([]);
});

it('fetches veBAL gauge as MAINNET', async () => {
    const vebalAddress = '0xE867AD0a48e8f815DC0cda2CDb275e0F163A480b';
    const repository = new VotingGaugesRepository();
    const rows = await repository.fetchOnchainVotingGauges([vebalAddress]);
    expect(rows).toEqual([
        {
            gaugeAddress: '0xe867ad0a48e8f815dc0cda2cdb275e0f163a480b',
            isInSubgraph: false,
            isKilled: true,
            network: 'MAINNET',
            relativeWeight: 0,
            relativeWeightCap: undefined,
        },
    ]);
});

export function aVotingGauge(...options: Partial<VotingGauge>[]): VotingGauge {
    const defaultRootGauge: VotingGauge = {
        gaugeAddress: '0x79ef6103a513951a3b25743db509e267685726b7',
        isKilled: false,
        network: 'MAINNET' as Chain,
        recipient: undefined,
        relativeWeight: 71123066693252456,
        relativeWeightCap: undefined,
        isInSubgraph: true,
    };
    return Object.assign({}, defaultRootGauge, ...options);
}

const EmptyError = new Error();

const repository = new VotingGaugesRepository(prismaMock);

it('successfully saves onchain gauges', async () => {
    const votingGauge = aVotingGauge({ network: Chain.OPTIMISM });

    const { votingGaugesWithStakingGaugeId: votingGauges } = await repository.saveVotingGauges([votingGauge]);

    expect(votingGauges[0]).toMatchObject(votingGauge);
    expect(votingGauges[0].stakingGaugeId).toBe(defaultStakingGaugeId);
});

describe('When staking gauge is not found ', () => {
    beforeEach(() => prismaMock.prismaPoolStakingGauge.findFirst.mockResolvedValue(null));

    it('has errors when gauge is valid for voting (not killed)', async () => {
        const repository = new VotingGaugesRepository(prismaMock);

        const votingGauge = aVotingGauge({ network: Chain.MAINNET, isKilled: false });

        const { votingGaugesWithStakingGaugeId, saveErrors } = await repository.saveVotingGauges([votingGauge]);

        expect(saveErrors.length).toBe(1);
    });

    it('does not throw when gauge is valid for voting (killed with no votes)', async () => {
        const votingGauge = aVotingGauge({ network: Chain.MAINNET, isKilled: true, relativeWeight: 0 });

        const result = await repository.saveVotingGauges([votingGauge]);

        expect(result).toBeDefined();
    });
});
