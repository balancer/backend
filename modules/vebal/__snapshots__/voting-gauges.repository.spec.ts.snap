// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VotingGaugesRepository > fetches veBAL gauge as MAINNET 1`] = `
[
  {
    "gaugeAddress": "0xe867ad0a48e8f815dc0cda2cdb275e0f163a480b",
    "isInSubgraph": false,
    "isKilled": "0",
    "network": "MAINNET",
    "relativeWeight": NaN,
    "relativeWeightCap": undefined,
  },
]
`;

exports[`VotingGaugesRepository > generates voting gauge rows given a list of gauge addresses 1`] = `
[
  {
    "gaugeAddress": "0x79ef6103a513951a3b25743db509e267685726b7",
    "isInSubgraph": false,
    "isKilled": "0.07308250729037725",
    "network": "MAINNET",
    "relativeWeight": NaN,
    "relativeWeightCap": undefined,
  },
  {
    "gaugeAddress": "0xfb0265841c49a6b19d70055e596b212b0da3f606",
    "isInSubgraph": false,
    "isKilled": "0",
    "network": "OPTIMISM",
    "relativeWeight": NaN,
    "relativeWeightCap": undefined,
  },
  {
    "gaugeAddress": "0x8f7a0f9cf545db78bf5120d3dbea7de9c6220c10",
    "isInSubgraph": false,
    "isKilled": "0",
    "network": "ARBITRUM",
    "relativeWeight": NaN,
    "relativeWeightCap": "0.000000000000000.02",
  },
]
`;
